src/pip/_internal/vcs/git.py

```diff
@@ -143,7 +152,8 @@ def resolve_revision(cls, dest, url, rev_options):
 
         if sha is not None:
             rev_options = rev_options.make_new(sha)
-            rev_options.branch_name = rev if is_branch else None
```

Comment:
Rather than changing this here in this way, it would probably be better to add `self.branch_name = None  # type: Optional[str]` inside `RevOptions.__init__()`.